archive/tar_test.go

```diff
@@ -96,82 +103,73 @@ func testTar(t *testing.T, when spec.G, it spec.S) {
 	})
 
 	when("#WriteTarArchive", func() {
-		it.Before(func() {
-			var err error
-			tmpDir, err = ioutil.TempDir("", "create-tar-test")
-			if err != nil {
-				t.Fatalf("failed to create tmp dir %s: %s", tmpDir, err)
-			}
-
-			tarFile = filepath.Join(tmpDir, "tar_test-go.tar")
-
-			file, err = os.Create(tarFile)
-			h.AssertNil(t, err)
-		})
-
-		it.After(func() {
-			err := os.RemoveAll(tmpDir)
-			h.AssertNil(t, err)
-
-			err = os.RemoveAll(tarFile)
-			h.AssertNil(t, err)
-		})
+		for _, src := range []string{
+			filepath.Join("testdata", "dir-to-tar"),
+			filepath.Join("testdata", "dir-to-tar") + string(filepath.Separator),
+			filepath.Join("testdata", "dir-to-tar") + string(filepath.Separator) + ".",
+		} {
+			src := src
+			it(fmt.Sprintf("writes a tar with the src filesystem contents (%s)", src), func() {
+				uid := 1234
+				gid := 4567
+
+				file, err := os.Create(filepath.Join(tmpDir, "tar_test-go.tar"))
+				h.AssertNil(t, err)
+				defer file.Close()
```

Comment:
We seem to be closing the same file multiple times (see line 121). It succeeds only b/c we aren't checking for errors here, which feels a bit messy.

Reply:
This is intentional. The defer is there for reassurance given that we might not actually reach the inlined `file.Close()` if something where to go wrong within `archive.WriteTarArchive`.

Reply:
If this was the implementation rather than a test I would suggest moving these few lines into a function and allowing the defer to close the file in both cases.

But given that this is test I am happy to be less nitpicky about it.